### PR TITLE
Feature mes 1728 remove old cached data

### PR DESCRIPTION
**Pull request by rossfauldsbjss**
Mes 1728 - remove old cached data.

amend app to remove data that is older than todays date minus _n_ configurable days.
This works for both journal and log data.  Logging had to be changed to work in offline mode and persist to storage as part of this.

Note: For journal data the data is the date the journal data was written. It will remove all the data once the date written is older than the cached days - regardless of whether the days within the journal data fall within the cache window.

## Pull Request checklist:

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist:

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop
- [ ] commits are squashed

## Sign off process checklist:

- [x] Code has been tested manually
- [x] Tests are passing
- [x] PR link added to JIRA
- [x] Reviewers selected in Github
- [ ] Any new reusable component/widget added to component library on Confluence

### CODE REVIEW

File: src/modules/logs/logs.actions.ts [hunk 1]
```diff
@@ -7,6 +7,10 @@ export const SEND_LOGS = '[LogsEffects] Send Logs';
 export const SEND_LOGS_SUCCESS = '[LogsEffects] Send Logs Success';
 export const SEND_LOGS_FAILURE = '[LogsEffects] Send Logs Failure';
 
+export const LOAD_LOG = '[GLOBAL] Load Logs';
+export const LOAD_LOG_STATE = '[GLOBAL] Load Log State';
+export const PERSIST_LOG = 'LogsEffects] Persist Logs';
```
Comment (ammarhaiderbjss) [inline src/modules/logs/logs.actions.ts]: missing an opening `[` bracket

Comment (rossfauldsbjss) [reply src/modules/logs/logs.actions.ts]: Added bracket

File: src/providers/journal/journal.ts [hunk 2]
```diff
@@ -47,13 +54,38 @@ export class JournalProvider {
...
+      };
+      this.dataStore.setItem('JOURNAL', JSON.stringify(journalDataToStore)).then((response) => {});
     }
   }
 
+  isCacheTooOld = (dateStored: DateTime, now: DateTime):boolean => {
+    return dateStored.daysDiff(now) > this.appConfigProvider.getAppConfig().daysToCacheJournalData;
+  }
+
+  emptyCachedData = () => {
```
Comment (ammarhaiderbjss) [inline src/providers/journal/journal.ts]: Ideally we should have a `journal.offline.effects.ts` where we house all of these dataStore interactions in. Much like how we do with the logs. In the absence of that, could we just add some docblocs to outline the purpose of these methods as related to the offline secure storage feature?

Comment (rossfauldsbjss) [reply src/providers/journal/journal.ts]: doc blocks added

File: src/providers/journal/__tests__/journal.spec.ts [hunk 3]
```diff
@@ -6,44 +6,117 @@ import { AuthenticationProviderMock } from '../../authentication/__mocks__/authe
...
+      };
+
+      // override mock getItem as we need data to test
+      // @ts-ignore
+      dataStoreMock.getItem.and.callFake(() => Promise.resolve(JSON.stringify(agedCache)));
+
+      spyOn(journalProvider, 'emptyCachedData').and.callThrough();
+
+      journalProvider.getAndConvertOfflineJournal().then((data) => {
+        expect(journalProvider.emptyCachedData).toHaveBeenCalled();
```
Comment (ammarhaiderbjss) [inline src/providers/journal/__tests__/journal.spec.ts]: Do these tests get called? We don't have `async` or `done()` being called here..

File: src/providers/journal/__tests__/journal.spec.ts [hunk 4]
```diff
@@ -6,44 +6,117 @@ import { AuthenticationProviderMock } from '../../authentication/__mocks__/authe
...
+      dataStoreMock.getItem.and.callFake(() => Promise.resolve(JSON.stringify(dataWthinWindowCache)));
 
-      httpMock = TestBed.get(HttpTestingController);
-      journalProvider = TestBed.get(JournalProvider);
-      authProviderMock = TestBed.get(AuthenticationProvider);
-      urlProviderMock = TestBed.get(UrlProvider);
+      spyOn(journalProvider, 'emptyCachedData').and.callThrough();
+
+      journalProvider.getAndConvertOfflineJournal().then((data) => {
+        expect(journalProvider.emptyCachedData).toHaveBeenCalledTimes(0);
```
Comment (ammarhaiderbjss) [inline src/providers/journal/__tests__/journal.spec.ts]: Same with these..

File: src/providers/journal/__tests__/journal.spec.ts (hunk 3 above)
Comment (ammarhaiderbjss) [reply src/providers/journal/__tests__/journal.spec.ts]: Ignore me, I think these are ok actually as they're wrapped in a `.then`

File: src/modules/logs/__tests__/logs.effects.spec.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+      // override mock getItem as we need data to test
+      // @ts-ignore
+      dataStoreMock.getItem.and.callFake(() => Promise.resolve(JSON.stringify(dataWthinWindowCache)));
+
+      spyOn(effects, 'emptyCachedData').and.callThrough();
+
+      effects.getAndConvertPersistedLogs().then((data) => {
+        expect(effects.emptyCachedData).toHaveBeenCalledTimes(0);
+        expect(dataStoreMock.setItem).toHaveBeenCalledTimes(0);
+        expect(data).toEqual(dataWthinWindowCache.data);
```
Comment (ammarhaiderbjss) [inline src/modules/logs/__tests__/logs.effects.spec.ts]: Just to be safe, could we please add some `done()` calls here at the end of these `.then` calls.

Comment (rossfauldsbjss) [reply src/modules/logs/__tests__/logs.effects.spec.ts]: done

**Pull request merged**